core/src/main/scala/kafka/log/LogCleanerManager.scala

```diff
@@ -345,16 +347,57 @@ private[log] object LogCleanerManager extends Logging {
     log.config.compact && log.config.delete
   }
 
+  /**
+    * return the baseOffset of the first dirty segment that was created at or after cleanUtilTime
+    */
+  def getMustCleanOffset(log: Log, firstDirtyOffset: Long, cleanUtilTime: Long) : Long = {
+
+    val largestCleanTimestamp = log.logSegments(-1, firstDirtyOffset).lastOption match {
+      case None => 0L
+      case Some(segment) => segment.largestTimestamp
+    }
+    val dirtyNonActiveSegments = log.logSegments(firstDirtyOffset, log.activeSegment.baseOffset)
+
+    val firstDirtySegmentCreateTime: Long = dirtyNonActiveSegments.headOption match {
+      case None => largestCleanTimestamp
+      case Some(segment) =>
+        val firstBatchTimestamp = segment.getFirstBatchTimestamp()
+        if (firstBatchTimestamp > 0)
+          firstBatchTimestamp
+        else {
+          math.max(largestCleanTimestamp, segment.largestTimestamp - segment.maxSegmentMs)
+        }
+    }
+
+    // find the baseOffset of the first dirty segment whose estimated creation time is greater or equal than cleanUtilTime
+    val dirtyNonActiveSegIter = dirtyNonActiveSegments.iterator
```

Comment:
I guess this is the core logic to calculate `mostCleanOffset`. It is better that engineers from confluent can help review this as well. 

For me, if I understand it correctly, iterator is used for performance reason to achieve O(n) look up, but I guess it could be simplified like:
```
    val greaterThanCleanUtilTime = dirtyNonActiveSegments.headOption.exists(_ => firstDirtySegmentCreateTime >= cleanUtilTime) ||
      dirtyNonActiveSegments.tail.exists(_.largestTimestamp >= cleanUtilTime)

    (log.firstUnstableOffset.map(_.messageOffset), greaterThanCleanUtilTime) match {
      case (None, true) => firstDirtyOffset
      case (None, false) => log.activeSegment.baseOffset
      case (Some(offset), true) => math.min(offset, firstDirtyOffset)
      case (Some(offset), false) => math.min(offset, log.activeSegment.baseOffset)
    }
```
- I am not sure `dirtyNonActiveSegments.exists(_.largestTimestamp >= cleanUtilTime)` is ok for this, but `headOption` and `tail` should achieve what you are trying to do.
- `exists` unlike `find` would return true/false instead of Option[Element], but both will end the iteration if the `condition` is matched. 
- I guess pattern matching makes more sense here, since List[Option[T]].flatten is hiding the logic to filter out None value implicitly.

Reply:
This doesn't work.  The must clean offset we are trying to find could be in the middle of dirty segments.  We have used fold and scan but iterator might be more clear for this purpose because we need to look at previous segment for current segment's creation time and look at next segment for baseoffset.

Reply:
I don't think this logic is a `fold`.  The fold operation would reduce the traversable to one target type `op: (A1, A1) => A1`. In this case `while (!foundOffset && dirtyNonActiveSegIter.hasNext)`, if `foundOffset ` returns true, the while loop will be ended and there would be no fold anymore. So the logic is more like `finding the possible mustCleanOffset`. We should probably use `exists` described above. :)

Reply:
I don't think exist approach will work here, because the offset we tried to find could be none of the four cases you mentioned :   firstDirtyOffset, log.activeSegment.baseOffset, log.firstUnstableOffset , firstDirtyOffset.  The offset could be in the middle of dirty segments, we need to look back one segment for creation time, and look forward one segment for baseoffset.

Although I don't like to use iterator,  it works in this case. 

say for example, you have 7 segments in the follow format (largesttimestamp, clean or dirty,  base offset)
(T1,  clean ,  0)   (T2, dirty,  10) (T3, dirty, 20) (T4, dirty, 30) (T5, dirty, 40) (T6, dirty, 50)  (T7, dirty active, 60)

if clean util time is  T3.5
we know the T4 segment is created on time T3(looking back), and it needs to be cleaned. 
and the must clean offset  is 40 (looking at next segment's base)